rust/worker/src/blockstore/arrow/sparse_index.rs

```diff
@@ -117,6 +117,209 @@ impl SparseIndex {
+            if let Some((next_key, _)) = next_iter.next() {
+                // This can't be a start key but we still need to extract it.
+                let non_start_next_key: Option<&CompositeKey>;
+                match next_key {
+                    SparseIndexDelimiter::Start => {
+                        panic!("Invariant violation. Sparse index is not valid.");
+                    }
+                    SparseIndexDelimiter::Key(k) => non_start_next_key = Some(k),
+                }
+                // If delimeter starts with the same prefix then there will be keys inside the
+                // block with this prefix.
+                if non_start_curr_key.is_some()
+                    && prefix == non_start_curr_key.unwrap().prefix.as_str()
+                {
+                    block_ids.push(*curr_uuid);
+                }
+                // If prefix is between the current delim and next delim then there could
+                // be keys in this block that have this prefix.
+                if (non_start_curr_key.is_none()
+                    || prefix > non_start_curr_key.unwrap().prefix.as_str())
+                    && (prefix <= non_start_next_key.unwrap().prefix.as_str())
+                {
+                    block_ids.push(*curr_uuid);
+                }
+            } else {
+                // Last block.
+                if non_start_curr_key.is_none()
+                    || prefix >= non_start_curr_key.unwrap().prefix.as_str()
+                {
+                    block_ids.push(*curr_uuid);
+                }
+            }
+        }
+        block_ids
+    }
+
+    pub(super) fn get_block_ids_gt<'a, K: ArrowReadableKey<'a> + Into<KeyWrapper>>(
+        &self,
+        prefix: &str,
+        key: K,
+    ) -> Vec<Uuid> {
+        let lock_guard = self.forward.lock();
+        let mut curr_iter = lock_guard.iter();
+        let mut next_iter = lock_guard.iter().skip(1);
+        let mut block_ids = vec![];
+        while let Some((curr_delim, curr_uuid)) = curr_iter.next() {
+            let curr_key = match curr_delim {
+                SparseIndexDelimiter::Start => None,
+                SparseIndexDelimiter::Key(k) => Some(k),
+            };
+            let mut next_key: Option<&CompositeKey> = None;
+            if let Some((next_delim, _)) = next_iter.next() {
+                next_key = match next_delim {
+                    SparseIndexDelimiter::Start => {
+                        panic!("Invariant violation. Sparse index is not valid.")
+                    }
+                    SparseIndexDelimiter::Key(k) => Some(k),
+                };
+            }
+            if curr_key.is_none() || curr_key.unwrap().prefix.as_str() < prefix {
+                if next_key.is_none() || next_key.unwrap().prefix.as_str() >= prefix {
+                    block_ids.push(*curr_uuid);
+                }
+            }
+            if curr_key.is_some() && curr_key.unwrap().prefix.as_str() == prefix {
```

Comment:
I can see why you wanted a thorough review on this. This looks correct to me.

Reply:
In hindsight this could be a good coding question to test ability to think through cases